### PR TITLE
Effort position controller fix

### PR DESCRIPTION
**Pull request by davetcoleman**
- Added enforeJointLimits() function
- Enforced joint limits at `update()` and `starting()` functions
- Cleaned up joint_state_controller
- Added .gitignore

This fixes a bug where if you start the controller and your current joint position is slightly outside the joint limits, the starting commanded position is also outside the limit, and the `shortest_angular_distance_with_limits` function returns an incorrect value of 2*PI instead of a small decimal. This causes the error to be very high and causes a lot of noise in the PID controller.

This bug is also issued here: https://github.com/ros/angles/issues/2

### CODE REVIEW

File: effort_controllers/src/joint_position_controller.cpp [hunk 1]
```diff
@@ -122,7 +122,15 @@ void JointPositionController::setCommand(double cmd)
 
 void JointPositionController::starting(const ros::Time& time)
 {
-  command_.initRT(joint_.getPosition());
+  ROS_DEBUG_STREAM_NAMED("joint_position_controller","Starting '" << joint_.getName()
```
Comment (adolfo-rt) [inline effort_controllers/src/joint_position_controller.cpp]: I'd rather not make log statements in code that will be executed in a realtime context. Rosconsole macros end up making system calls which trigger a switch from primary (realtime) to secondary (linux) mode.

File: effort_controllers/src/joint_position_controller.cpp [hunk 2]
```diff
@@ -189,6 +202,24 @@ void JointPositionController::setCommandCB(const std_msgs::Float64ConstPtr& msg)
...
 
+void JointPositionController::enforceJointLimits(double &command)
+{
+  // Check that this joint has applicable limits
+  if (joint_urdf_->type == urdf::Joint::REVOLUTE || joint_urdf_->type == urdf::Joint::PRISMATIC)
+  {
+    if( command > joint_urdf_->limits->upper ) // above upper limnit
+    {
+      command = joint_urdf_->limits->upper;
+      ROS_WARN_STREAM_NAMED("temp","Command value is greater than upper limit");
```
Comment (adolfo-rt) [inline effort_controllers/src/joint_position_controller.cpp]: Ditto here and below warning. Printing these messages will trigger a mode switch on a realtime OS. One can argue that these should be rare occurrences, but I'd rather document clearly that this controller saturates commands and omit the log statements.

File: effort_controllers/src/joint_position_controller.cpp (hunk 1 above)
Comment (davetcoleman) [inline effort_controllers/src/joint_position_controller.cpp]: I'll happily remove this but isn't it ok for the constructor and this `starting` function to be non-realtime?

Comment (adolfo-rt) [inline effort_controllers/src/joint_position_controller.cpp]: In the controller manager, starting and stopping a controller is done inside the realtime [update()](https://github.com/ros-controls/ros_control/blob/hydro-devel/controller_manager/src/controller_manager.cpp#L83-L84) method. Controller construction and `init()`ialization is non-realtime ,though.

I just checked that the [ControllerBase](https://github.com/ros-controls/ros_control/blob/hydro-devel/controller_interface/include/controller_interface/controller_base.h) class documentation is not using the `\name` blocks present in other parts of ros_control. That, at least, can be improved:

``` c++
/** \name Real-Time Safe Functions
*\{*/
...
/*\}*/


/** \name Non Real-Time Safe Functions
*\{*/
...
/*\}*/
```

Comment (adolfo-rt) [inline effort_controllers/src/joint_position_controller.cpp]: Improved ControllerBase documentation [here]((https://github.com/pal-robotics/ros_control/commit/4e23e2c4d7f1c55953d142c712aa046f6715185c).

Comment (davetcoleman) [inline effort_controllers/src/joint_position_controller.cpp]: Removed the debug output.

**Pull request merged**